### PR TITLE
Fix memory leak for ColumnFamily drop with live iterator

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: Uncommon bug seen by ASAN with
ColumnFamilyTest.LiveIteratorWithDroppedColumnFamily, if the last two
references to a ColumnFamilyData are both SuperVersions (during
InstallSuperVersion). The fix is to use UnrefAndTryDelete even in
SuperVersion::Cleanup but with a parameter to avoid re-entering Cleanup
on the same SuperVersion being cleaned up.

ColumnFamilyData::Unref is considered unsafe so removed.

Test Plan: ./column_family_test --gtest_filter=*LiveIter* --gtest_repeat=100

### CODE REVIEW

**Comment by pdillinger** [comment]
> Hi @pdillinger 👋
> Getting in touch on behalf of Snyk. Could this memory leak represent a security risk? I.e. could a malicious user practically exploit it and trigger a DoS on the victim server?

Dropping a column family is not normally done in response to service user data, but for maintenance or internal data migrations.

File: db/column_family.cc [hunk 1]
```diff
@@ -449,9 +449,7 @@ void SuperVersion::Cleanup() {
     to_delete.push_back(m);
   }
   current->Unref();
-  if (cfd->Unref()) {
-    delete cfd;
-  }
+  cfd->UnrefAndTryDelete(this);
```
Comment (jay-zhuang) [inline db/column_family.cc:452]: Just a question: when does `ColumnFamilyData` have different `SuperVersion`?
I thought they're always point to each other for the same CF.

File: db/column_family.cc [hunk 2]
```diff
@@ -670,7 +668,11 @@ bool ColumnFamilyData::UnrefAndTryDelete() {
     return true;
   }
 
-  if (old_refs == 2 && super_version_ != nullptr) {
+  // If called under SuperVersion::Cleanup, we should not re-enter Cleanup on
+  // the same SuperVersion. (But while installing a new SuperVersion, this
+  // cfd could be referenced only by two SuperVersions.)
+  if (old_refs == 2 && super_version_ != nullptr &&
+      super_version_ != sv_under_cleanup) {
```
Comment (jay-zhuang) [inline db/column_family.cc:675]: If `sv_under_cleanup` is only used by `SuperVersion::Cleanup()`, should this logic also move to there? I feel having a SuperVersion a parameter for `UnrefAndTryDelete` is a little confusing.

File: db/column_family.cc (hunk 1 above)
Comment (pdillinger) [reply db/column_family.cc:452]: From new comment

    // (But while installing a new SuperVersion, this
    // cfd could be referenced only by two SuperVersions.)

File: db/column_family.cc (hunk 2 above)
Comment (pdillinger) [reply db/column_family.cc:675]: When sv_under_cleanup is non-null, we might exclude this code. We might need the code below from any caller. If you have a specific proposal to simplify, let me know. It is not simple as it is here.

Comment (jay-zhuang) [reply db/column_family.cc:675]: I think my suggestion is to change `SuperVersion::Cleanup()`, like:
```
  if (cfd->GetSuperVersion() == this) {
    if (cfd->Unref()) {
      delete cfd;
    }
  } else {
    cfd->UnrefAndTryDelete();
  }
```

**Review by jay-zhuang — APPROVED** [review]
LGTM.

File: db/column_family.cc (hunk 1 above)
Comment (jay-zhuang) [reply db/column_family.cc:452]: oh, I see, one is the new SV, another one is the old SV.

File: db/column_family.cc (hunk 2 above)
Comment (pdillinger) [reply db/column_family.cc:675]: We would either need to keep the Unref() function around, which I fear invites further bugs, or make SuperVersion a friend of ColumnFamilyData. Making a `friend` suggests we would be violating the natural encapsulation for ColumnFamilyData.

**Pull request closed without merging**